prefect_shell/commands.py

```diff
@@ -110,3 +121,95 @@ def example_shell_run_command_flow():
 
     line = lines[-1] if lines else ""
     return lines if return_all else line
+
+
+class ShellJobRun(JobRun):
+    """
+    A class representing a run of a shell job.
+    """
+
+    def __init__(self, shell_job: "ShellJob", process: Process):
+        self._shell_job = shell_job
+        self._process = process
+        self._output = []
+
+    async def _capture_output(self, source):
+        """
+        Capture output from source.
+        """
+        async for text in TextReceiveStream(source):
+            if self._shell_job.stream_output:
+                self.logger.info(text)
+            self._output.append(text)
+
+    @sync_compatible
+    async def wait_for_completion(self):
+        """
+        Wait for the job run to complete.
+        """
+        await asyncio.gather(
+            self._capture_output(self._process.stdout),
+            self._capture_output(self._process.stderr),
+        )
+
+    @sync_compatible
+    async def fetch_result(self) -> List[str]:
+        """
+        Retrieve the results of the job run and return them.
+        """
+        return self._output
+
+
+class ShellJob(JobBlock):
```

Comment:
I think we need a different name for this. Maybe something like `ShellOperation`?

Also, I think this block could feasibly have a method that handles all three steps in one call, but there's probably multiple ways we could expose that functionality.